### DISCUSSION TITLE
Mount the user's SSH config file to Support 1Password SSH agent and possibly other agents

### DISCUSSION THREAD

**Discussion by nckrtl** [serversideup/spin#81] [Ideas]
It would be nice if the user's ssh config would be mounted as well by default when running the provision command. And perhaps pass it by default. Typically the ssh config file is located at `~/.ssh/config`.

Currently you can only pass a config file by copying the ssh config file manually, add it to the root of the project and run `spin provision -u root --ssh-extra-args='-F config'`

By mounting/passing the ssh config file support can be added for custom ssh agents which can be configured through the ssh config file, like 1Password's ssh agent.

**Comment by jaydrogers** [comment]
Thanks! I will take a look into this and will keep you posted 👍

**Comment by jaydrogers** [comment]
I finally had a chance to take a look into this. I am mounting the entire .ssh folder here:

https://github.com/serversideup/spin/blob/62c26d4fd2a2148123ea687c6516e478a60d4d55/lib/functions.sh#L701-L708

I'm assuming this should include the `.ssh/config` file automatically.

Is it something with Linux not being able to interpret the 1password integration?

We're running Ansible & SSH from within the Ansible container (not macOS)

**Reply by nckrtl** [reply]
In that case I'll take another look at it. At the time I'm pretty sure it couldn't find the config file. Maybe I overlooked something. Will be on holiday this week so will take a look at it next week.

**Reply by jaydrogers** [reply]
No problem! Enjoy your holiday break!!

**Reply by nckrtl** [reply]
Hmm, cant seem to get it to work without adding the config file in the project root where I run the provision command "--ssh-extra-args='-F config'". Also the "--ssh-extra-args" flag caused the provision to get stuck/not start at all. I notice the provision.sh content has changed recently. 

The previous version of provision.sh (still used in this branch: https://github.com/serversideup/spin/blob/75-spin-deploy-allow-deployments-without-cicd/lib/actions/provision.sh) does process it.

I think ansible would need additional configuration to also process ssh config file if any is available. Not very familiar with Ansible but there seems to be a specific module: https://docs.ansible.com/ansible/latest/collections/community/general/ssh_config_module.html

**Reply by nckrtl** [reply]
Been a bit side-tracked with client work but but I gave it another look. I can confirm the ssh config file is available within the docker container. For example, named connections specified in the ssh config can be used within .spin-inventory.ini instead of using the ip address.

After trying it again it seems to properly forward the ssh agent but only when specifying `ssh-extra-args='-F <any-directory-name>`. When I actually specify the ssh config file (which I pasted into the project) it doesn't work. When specifiying for example the path where the config file is copied to (/root/.ssh/config) it also doesn't work. It only works when specific a random directory name. So it seems `ssh-extra-args='-F <any-directory-name>` triggers something that allows to reach/trigger the 1Password ssh agent from the docker container.

So not sure yet what the root cause is, but thought it would be helpful to share these findings.

**Reply by jaydrogers** [reply]
Hey @nckrtl,

I am FINALLY catching up to this too 😆

Could you try this again with Spin v2.2.0 when you have a chance?

I made a ton of improvements to SSH and I am hoping this resolves it. I don't have 1password on my machine, otherwise I would've tested this myself:

https://github.com/serversideup/spin/releases/tag/v2.2.0

**Reply by nckrtl** [reply]
I'm getting ansible errors with 2.2.0 when deploying/provisioning as described here https://github.com/serversideup/spin/issues/115. Should work out of the box right?

**Reply by nckrtl** [reply]
Still same behaviour, but now my workaround is also no longer working. But ssh-extra-args='-F <any-directory-name>' seems to actually look for the file specified, as it must be a valid path else I get an error the config isn't found. Before this was not the case.

I'm not surprised the workaround no longer works btw as it felt pretty hacky that is was working anyways. I will need to figure out the proper way to communicate with 1Password I guess.

**Reply by nckrtl** [reply]
After some more digging it turns that RUN_AS_USER is causing issues for me in run_ansible() in functions.sh when provisioning

When I change it from
```
  docker run --rm -it \
    -e "PUID=${SPIN_USER_ID}" \
    -e "PGID=${SPIN_GROUP_ID}" \
    -e "RUN_AS_USER=$(whoami)" \
    "${additional_docker_args[@]}" \
    "$SPIN_ANSIBLE_IMAGE" \
    "${args_without_options[@]}"
 ```
 
 to: (also removed PUID and GUID as it caused issues as well described here: https://github.com/serversideup/spin/issues/115#issuecomment-2401885888)
 ```
  docker run --rm -it \
    "${additional_docker_args[@]}" \
    "$SPIN_ANSIBLE_IMAGE" \
    "${args_without_options[@]}"
 ```
 
 It actually does forward the ssh connection to 1password

**Reply by jaydrogers** [reply]
Working on a solution. I think I have it fixed, I just have to push a release. Stay tuned 😃

**Reply by jaydrogers** [reply]
This might fix it 🤞

https://github.com/serversideup/spin/issues/115#issuecomment-2402821143

**Comment by nckrtl** [comment]
Jay got it resolved, check out the solution here: https://serversideup.net/how-to-get-ssh-to-work-with-1password-docker-desktop-macos-within-a-container/
